modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -4933,6 +4951,14 @@ private void processNodeAddFinishedMessage(TcpDiscoveryNodeAddFinishedMessage ms
                     mux.notifyAll();
                 }
 
+                if (gridDiscoveryData != null)
+                    spi.onExchange(gridDiscoveryData, U.resolveClassLoader(spi.ignite().configuration()));
+
+                for (DiscoveryDataPacket dataPacket : joiningNodesDiscoDataList)
```

Comment:
It's not clear that joiningNodesDiscoDataList will not be null in this piece of code, could you please leave a comment about it or provide explicit null check?

Reply:
Indeed there are execution branches leaving this collection uninitialized. Although I haven't seen NPE at this line anywhere in our tests it is safer to add explicit null check.